### PR TITLE
Migrate to Typescript

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
I have attempted to follow the lonlat transition "style" as closely as possible. Package's tests pass as do OTP-UI's when using this package.

@landonreed @evansiroky

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,32 +13,29 @@
...
+    "test": "yarn run lint && yarn run lint-docs && yarn jest",
+    "test-node": "yarn build && tsdx test --test-environment node"
   },
   "dependencies": {
-    "@conveyal/lonlat": "^1.3.0",
-    "geocoder-arcgis": "^2.0.4"
+    "@conveyal/lonlat": "^1.4.1",
+    "eslint-config-standard-jsx": "^10.0.0",
+    "geocoder-arcgis": "^2.0.5",
+    "tsdx": "https://github.com/ibi-group/tsdx"
```
Comment (evansiroky) [inline package.json:31]: This should be a devDependency.

File: geocoder-arcgis.d.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
+declare class ArcGISAuth {
```
Comment (evansiroky) [inline geocoder-arcgis.d.ts:1]: Is this file needed?

File: index.ts [hunk 3]
```diff
@@ -1,30 +1,62 @@
...
+import { toCoordinates, toString, toPoint } from '@conveyal/lonlat'
+import type { LonLatInput } from '@conveyal/lonlat'
 
-const arcGisGeocoders = {}
+type Point = {
+  x: number
+  y: number
+}
+
+type GeocodersList = {
```
Comment (evansiroky) [inline index.ts:11]: Nit suggestion:

```suggestion
type GeocodersMap = {
```

File: index.ts [hunk 4]
```diff
@@ -1,30 +1,62 @@
...
+  forStorage?: boolean
+  outFields?: string
+  magicKey?: string
+  size?: number
+}
+
 type BaseQuery = {
-  clientId?: string,
-  clientSecret?: string,
+  boundary?: Boundary
```
Comment (evansiroky) [inline index.ts:39]: Please remove the `boundary` key as it is not used in the `reverse` function.

File: index.ts [hunk 5]
```diff
@@ -1,30 +1,62 @@
...
+  clientId?: string
+  clientSecret?: string
+  focusPoint?: Point
+  text?: string
```
Comment (evansiroky) [inline index.ts:43]: Please remove `boundary`, `focusPoint` and `text` since they are not used in every public function call.

File: index.ts [hunk 6]
```diff
@@ -1,30 +1,62 @@
...
+}
+
+type Options = {
+  searchExtent?: string
+  location?: string
+  forStorage?: boolean
+  outFields?: string
+  magicKey?: string
+  size?: number
+}
```
Comment (evansiroky) [inline index.ts:36]: I would omit this type entirely because it might produce a type definition for a public function where some or all of these are not used by a given function.

File: index.ts [hunk 7]
```diff
@@ -113,40 +152,35 @@ function candidateToGeojson (candidate) {
...
   text,
   url
-}: BaseQuery & {
-  boundary?: Boundary,
-  focusPoint?: any,
-  text: string,
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  const options = {}
+}: BaseQuery & Options): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:162]: Two things:

1. Can you in-line the applicable keys within the `Options` type since not all of those values would apply as keys in the input argument for this function call?
2. Can you also please extend the `BaseResponse` type with the `query` key/value since the query is eventually returned?

File: index.ts [hunk 8]
```diff
@@ -163,34 +197,32 @@ export function autocomplete ({
...
-  addresses: Array<any>,
-  boundary?: Boundary,
-  focusPoint?: any
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  let addressesQuery = [...addresses]
-  const options = {}
+}: BaseQuery & Options & { addresses: Array<string> }): Promise<
+  BaseResponse[]
+> {
```
Comment (evansiroky) [inline index.ts:209]: Two things:

1. Can you in-line the applicable keys within the `Options` type since not all of those values would apply as keys in the input argument for this function call?
2. Is the `Promise<BaseResponse[]>` return type correct? It seems it may be more like `Promise<BaseResponse & { query: Array<String> }>`.

File: index.ts [hunk 9]
```diff
@@ -229,45 +262,48 @@ export function bulk ({
...
   point,
   url
-}: BaseQuery & {
-  forStorage?: boolean,
-  point: any
-}): Promise<BaseResponse> {
+}: BaseQuery &
+  Options & {
+    point: LonLatInput
+  }): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:274]: Two things:

1. Can you in-line the applicable keys within the `Options` type since not all of those values would apply as keys in the input argument for this function call?
2. Can you also please extend the `BaseResponse` type with the `query` key/value since the query is eventually returned?

File: index.ts [hunk 10]
```diff
@@ -289,34 +325,27 @@ export function reverse ({
...
-  boundary?: Boundary,
-  focusPoint?: any,
-  forStorage?: boolean,
-  magicKey?: string,
-  size?: number,
-  text: string,
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  const options = {}
+}: BaseQuery & Options): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:338]: Two things:

1. Can you in-line the applicable keys within the `Options` type since not all of those values would apply as keys in the input argument for this function call?
2. Can you also please extend the `BaseResponse` type with the `query` key/value since the query is eventually returned?

File: package.json [hunk 11]
```diff
@@ -11,32 +13,29 @@
...
+    "build": "tsdx build -i index.ts --target node",
+    "cover": "tsdx test --coverage --coverage-paths index.ts",
+    "generate-docs": "documentation readme index.ts --document-exported=true -g --section=API",
+    "jest": "yarn build && tsdx test",
+    "lint": "tsdx lint __tests__ index.ts --quiet",
+    "lint-docs": "documentation lint index.ts",
+    "prepublish": "yarn build",
+    "semantic-release": "semantic-release",
+    "test": "yarn run lint && yarn run lint-docs && yarn jest",
+    "test-node": "yarn build && tsdx test --test-environment node"
```
Comment (evansiroky) [inline package.json:25]: Is this still needed?

**Review by evansiroky — CHANGES_REQUESTED** [review]
Nice work so far! Just a few comments.

File: geocoder-arcgis.d.ts (hunk 2 above)
Comment (miles-grant-ibigroup) [reply geocoder-arcgis.d.ts]: I'm surprised it's not! This types for the `geocoder-arcgis` dependency, whose types are used quite often throughout the code...

File: package.json (hunk 11 above)
Comment (miles-grant-ibigroup) [reply package.json:25]: [It's used in the CI](https://github.com/miles-grant-ibigroup/geocoder-arcgis-geojson/blob/c0b51cdc2ccce186ed6109d089331898c18e81d6/.github/workflows/node-ci.yml#L27). Not sure if `yarn test` would work just as well, but didn't want to create more changes than needed. You think it's worth changing?

**Comment by miles-grant-ibigroup** [comment]
Thanks! Have made the changes. Just a few questions about some specifics.

File: package.json (hunk 11 above)
Comment (evansiroky) [reply package.json:25]: I see how it works now in CI, we can keep as is.

File: index.ts [hunk 12]
```diff
@@ -229,45 +259,47 @@ export function bulk ({
...
   clientId,
   clientSecret,
   forStorage = false,
   point,
   url
-}: BaseQuery & {
-  forStorage?: boolean,
-  point: any
+}: BaseQuery & { forStorage?: boolean } & {
+  point: LonLatInput
```
Comment (evansiroky) [inline index.ts:269]: Can the two extra typed objects after BaseQuery be combined into one?

File: index.ts [hunk 13]
```diff
@@ -1,30 +1,50 @@
...
 
-type BaseResponse = {
-  query: any
+type GeocoderResponse = {
+  suggestions?: string[]
+  locations?: Candidate[]
+  location?: LonLatInput
+  address?: Record<string, string>
+  candidates?: Candidate[]
 }
```
Comment (evansiroky) [inline index.ts:41]: This is too generic of a type to use for the parse results of each geocoding response. I recommend creating a specific response type for each API call (`AutocompleteResponse`, `BulkResopsne`, `ReverseResponse`, `SearchResponse`).

File: index.ts [hunk 14]
```diff
@@ -1,30 +1,50 @@
...
 
-function boundaryToSearchExtent (boundary: Boundary): string {
+type BaseResponse = GeoJSON.FeatureCollection & { query: Array<string> }
```
Comment (evansiroky) [inline index.ts:43]: This `BaseResponse` type doesn't work in every API call's return type. Each is specific enough that I don't think it's worth defining an overall type here.

File: index.ts [hunk 15]
```diff
@@ -113,40 +140,39 @@ function candidateToGeojson (candidate) {
...
   text,
   url
 }: BaseQuery & {
-  boundary?: Boundary,
-  focusPoint?: any,
-  text: string,
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
 }): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:154]: The `BaseResponse` is not even correct here as the query part is an Object with the text key/value. Please update the return type.

File: index.ts [hunk 16]
```diff
@@ -163,48 +189,52 @@ export function autocomplete ({
...
   url
 }: BaseQuery & {
-  addresses: Array<any>,
-  boundary?: Boundary,
-  focusPoint?: any
+  addresses: Array<string>
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
 }): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:204]: The `BaseResponse` is not even correct here as the query part is an Object with the addresses key/value. Please update the return type.

File: index.ts [hunk 17]
```diff
@@ -229,45 +259,47 @@ export function bulk ({
...
   clientSecret,
   forStorage = false,
   point,
   url
-}: BaseQuery & {
-  forStorage?: boolean,
-  point: any
+}: BaseQuery & { forStorage?: boolean } & {
+  point: LonLatInput
 }): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:270]: The `BaseResponse` is not even correct here as the query part is a lonlat type. Please update the return type.

File: index.ts [hunk 18]
```diff
@@ -289,34 +321,41 @@ export function reverse ({
...
-  magicKey?: string,
-  size?: number,
-  text: string,
+  boundary?: Boundary
+  focusPoint?: Point
+  forStorage?: boolean
+  magicKey?: string
+  size?: number
+  text?: string
 }): Promise<BaseResponse> {
```
Comment (evansiroky) [inline index.ts:341]: The `BaseResponse` is not even correct here as the query part is an Object with the text key/value. Please update the return type.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Some more typing changes requested.

File: index.ts (hunk 12 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Yes they can! In cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 13 above)
Comment (miles-grant-ibigroup) [reply index.ts]: All of them were inline-able! Good recommendation thank you. Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 14 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 15 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 16 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 17 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts (hunk 18 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Fixed in cdd90839c90a85888732fb6718964b8ddd8fb1b6

File: index.ts [hunk 19]
```diff
@@ -113,32 +130,33 @@ function candidateToGeojson (candidate) {
...
-  boundary?: Boundary,
-  focusPoint?: any,
-  text: string,
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  const options = {}
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
+}): Promise<GeoJSON.FeatureCollection & { query: { text: string } }> {
```
Comment (evansiroky) [inline index.ts:144]: This does not actually return GeoJSON. It returns the data nearly verbatim from the [suggest](https://developers.arcgis.com/rest/geocode/api-reference/geocoding-suggest.htm) API endpoint. I would write this type as follows:

```suggestion
}): Promise<{
  features: Array<SuggestOption>
  query: { text: string }
}> {
```

where `SuggestOption` is defined as:

```js
type SuggestOption = {
  isCollection: boolean
  magicKey: string
  text: string
}
```

File: index.ts [hunk 20]
```diff
@@ -113,32 +130,33 @@ function candidateToGeojson (candidate) {
...
   if (boundary) {
     options.searchExtent = boundaryToSearchExtent(boundary)
   }
 
   // make request to arcgis
-  return geocoder.suggest(text, options)
-    .then(response => {
+  return geocoder
+    .suggest(text, options)
+    .then((response: { suggestions: string[] }) => {
```
Comment (evansiroky) [inline index.ts:159]: Please use the recommended type in combination with the above comment:

```suggestion
    .then((response: { suggestions: SuggestOption[] }) => {
```

File: index.ts [hunk 21]
```diff
@@ -163,48 +181,52 @@ export function autocomplete ({
...
-  focusPoint?: any
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  let addressesQuery = [...addresses]
-  const options = {}
+  addresses: Array<string>
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
+}): Promise<GeoJSON.FeatureCollection> {
```
Comment (evansiroky) [inline index.ts:196]: This is missing the query part of the response. See recommended change:

```suggestion
}): Promise<GeoJSON.FeatureCollection & { query: { addresses: String[] } }> {
```

File: index.ts [hunk 22]
```diff
@@ -1,30 +1,40 @@
...
+    minLon: number
+    maxLat: number
     maxLon: number
   }
 }
 
+type Candidate = {
+  location: LonLatInput
+  attributes: Record<string, string | number>
+}
```
Comment (evansiroky) [inline index.ts:27]: It's probably fine to not spell out every exact attribute here, but it would be nice to include some additional comments. Also, the location key while parseable by lonlat, has a more specific type. Recommended change:

```suggestion
/**
 * A geocoding candidate as received from either the findAddressCandidates or geocodeAddresses endpoint.
 * See info about some attributes here: https://developers.arcgis.com/rest/geocode/api-reference/geocoding-service-output.htm
 */
type Candidate = {
  location: {
    x: number
    y: number
  }
  attributes: Record<string, string | number>
}
```

**Review by evansiroky — CHANGES_REQUESTED** [review]
Getting closer. I'm just trying to make sure we get these typings right.

File: index.ts (hunk 19 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Refactored in aecf758d679984328a465df451fd953bde17efbf

File: index.ts (hunk 20 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Refactored in aecf758d679984328a465df451fd953bde17efbf

File: index.ts (hunk 21 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Refactored in aecf758d679984328a465df451fd953bde17efbf

File: index.ts (hunk 22 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Thanks for the suggestion! Added in aecf758d679984328a465df451fd953bde17efbf

**Comment by miles-grant-ibigroup** [comment]
Thanks so much! Your knowledge of the library is very helpful.

File: index.ts [hunk 23]
```diff
@@ -113,32 +142,33 @@ function candidateToGeojson (candidate) {
...
-  boundary?: Boundary,
-  focusPoint?: any,
-  text: string,
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  const options = {}
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
+}): Promise<Array<SuggestOption> & { query: { text: string } }> {
```
Comment (evansiroky) [inline index.ts:156]: This is missing the `features` key.

```suggestion
}): Promise<{
  features: Array<SuggestOption>
  query: { text: string } 
}> {
```

**Review by evansiroky — CHANGES_REQUESTED** [review]
One last thing, then should be good to go.

File: index.ts (hunk 23 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Have fixed all endpoints in 084c5d3b0eaa3897e885428831db509ede8e5391

**Comment by miles-grant-ibigroup** [comment]
Thanks for catching that last type error, I think it applied to all methods!

**Comment by evansiroky** [comment]
> Thanks for catching that last type error, I think it applied to all methods!

Actually, it just applied to the `autoComplete` function. The `GeoJSON.FeatureCollection` type is an object type with the `features` key. See here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/18b0118c4326e9f7e14221a96faeba3d65e2d2bd/types/geojson/index.d.ts#L172-L175

File: index.ts [hunk 24]
```diff
@@ -163,48 +193,55 @@ export function autocomplete ({
...
-}): Promise<BaseResponse> {
-  const geocoder = getGeocoder(clientId, clientSecret, url)
-  let addressesQuery = [...addresses]
-  const options = {}
+  addresses: Array<string>
+  boundary?: Boundary
+  focusPoint?: Point
+  text?: string
+}): Promise<{
+  features: GeoJSON.FeatureCollection
```
Comment (evansiroky) [inline index.ts:209]: This is not the correct type since `GeoJSON.FeatureCollection` consists of an object with the `features` key.

File: index.ts [hunk 25]
```diff
@@ -229,47 +266,55 @@ export function bulk ({
...
   forStorage = false,
   point,
   url
 }: BaseQuery & {
-  forStorage?: boolean,
-  point: any
-}): Promise<BaseResponse> {
+  forStorage?: boolean
+  point: LonLatInput
+}): Promise<{ features: GeoJSON.FeatureCollection; query: Point }> {
```
Comment (evansiroky) [inline index.ts:278]: This is not the correct type since `GeoJSON.FeatureCollection` consists of an object with the `features` key.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See latest comments.

File: index.ts (hunk 24 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Ah thank you! Have added a comment about this as well in 8251d96fcd55c1a47faa4ece03b7a53c7122f955

File: index.ts (hunk 25 above)
Comment (miles-grant-ibigroup) [reply index.ts]: Ah thank you! Have added a comment about this as well in 8251d96fcd55c1a47faa4ece03b7a53c7122f955

**Comment by miles-grant-ibigroup** [comment]
Ah thank you for clearing this up! Both have been fixed.

**Review by evansiroky — APPROVED** [review]

File: __tests__/index.test.ts [hunk 26]
```diff
@@ -1,8 +1,8 @@
 // @flow
 
-import nock from 'nock'
+import * as nock from 'nock'
 
-import {autocomplete, bulk, reverse, search} from '../index'
+import { autocomplete, bulk, reverse, search } from '../dist/index'
```
Comment (landonreed) [inline __tests__/index.test.ts:5]: I don't fully get why we import from dist here. Could a comment help explain this?

**Review by landonreed — APPROVED** [review]
Looks good! Just that one question about the import from `/dist`

File: __tests__/index.test.ts (hunk 26 above)
Comment (miles-grant-ibigroup) [reply __tests__/index.test.ts:6]: Good call. Comment added in 503172a

**Comment by miles-grant-ibigroup** [comment]
Clarification added thank you @landonreed! Assigning back to both of you for merge.

**Comment by evansiroky** [comment]
> Clarification added thank you @landonreed! Assigning back to both of you for merge.

It isn't that jest can't do this, it's more that the transformation isn't configured correctly. Using https://github.com/ibi-group/tsdx/pull/2 somewhat helps, but the coverage side of things is still broken. Regardless, I'm fine with merging as is for now to be done.
